Cognitive Services 2024-04-01-preview

tianxin-ms opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/20296335/9e51cff7-28f7-4ae7-99a3-0f4a116c759f)

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

tianxin-ms committed:
Adds base for updating Microsoft.CognitiveServices from version preview/2023-10-01-preview to version 2024-04-01-preview

tianxin-ms committed:
Updates readme

tianxin-ms committed:
Updates API version in new specs and examples

tianxin-ms committed:
Cognitive Services 2024-04-01-preview

tianxin-ms committed:
Add NSP APIs

tianxin-ms committed:
address some error

tianxin-ms committed:
address lintdiff error

tianxin-ms committed:
Add list for defender

tianxin-ms committed:
fix nsp ModelValidation

tianxin-ms committed:
Add Mode Capacities Apis

tianxin-ms committed:
fix example

tianxin-ms committed:
Added properties for models api

tianxin-ms committed:
Update models APIs

tianxin-ms commented:
LintDiff issue: ResourceNameRestriction, AvoidAdditionalProperties are from old version.

tianxin-ms committed:
fix a name

tianxin-ms committed:
Removed getModel

tianxin-ms committed:
Change an example name

tianxin-ms committed:
Fix example name

tianxin-ms committed:
Update NSP APIs

raosuhas commented:
>               "$ref": "#/definitions/ModelListResult"

[ARMBlockingComment] you also need to support individual gets on each of these Model resources

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:915 in 5a29fea. [](commit_id = 5a29fea34bf9f8e7436a7f52dae7fa8c82000029, deletion_comment = False)

raosuhas commented:
>       "post": {

Why do you need a separate POST action to update the block list items ? 

Ifthe updates are part of the body then you should be able to model it as PUT\PATCH , else you can child resources for items instead.

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:2665 in 5a29fea. [](commit_id = 5a29fea34bf9f8e7436a7f52dae7fa8c82000029, deletion_comment = False)

raosuhas commented:
>           "description": "The location of the Model."

use the tracked resources definition from common types for this. 

Also if this has a location then this is artacked resource , so you need to support PUT \ DELETE as well

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:5730 in 5a29fea. [](commit_id = 5a29fea34bf9f8e7436a7f52dae7fa8c82000029, deletion_comment = False)

raosuhas commented:
>         },

use tracked resource from commontypes

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:5786 in 5a29fea. [](commit_id = 5a29fea34bf9f8e7436a7f52dae7fa8c82000029, deletion_comment = False)

raosuhas commented:
> LintDiff issue: ResourceNameRestriction, AvoidAdditionalProperties are from old version.

Could you fix the warnings introduced in the new version. Specfically you should use v5 of the common types

tianxin-ms commented:
> > ```
> >   "post": {
> > ```
> 
> Why do you need a separate POST action to update the block list items ?
> 
> Ifthe updates are part of the body then you should be able to model it as PUT\PATCH , else you can child resources for items instead.
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:2665 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)

These 2 new RaiBlocklistItems APIs are for batch operation (add and delete)

xiaomin-ms commented:
> > ```
> >           "$ref": "#/definitions/ModelListResult"
> > ```
> 
> [ARMBlockingComment] you also need to support individual gets on each of these Model resources
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:915 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)

Suhas, we would like your suggestions here. Here is the full story, we originally has a "locations/models" API which return a list of models with self-defined metadata style schema: no "id", "type", "name" and "properties", from this version, we are adding a subscription wide "models" API which use standard ARM resource schema with "id", "type", "name" and "properties". We did add get by Id originally with the new schema, but the tool complain that the schema must be the same as list location models API, that's why we skip adding get by id.

tianxin-ms committed:
update PutRaiPolicy example

raosuhas commented:
> > > ```
> > >   "post": {
> > > ```
> > 
> > 
> > Why do you need a separate POST action to update the block list items ?
> > Ifthe updates are part of the body then you should be able to model it as PUT\PATCH , else you can child resources for items instead.
> > Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:2665 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)
> 
> These 2 new RaiBlocklistItems APIs are for batch operation (add and delete)

In ARM we always prefer to model things as resources even items since all tooks like templates \ policy etc depend on the items being modelled as resources. SO for example if you model this as a POST you will not be able to create this as part of a template.

raosuhas commented:
> > > ```
> > >           "$ref": "#/definitions/ModelListResult"
> > > ```
> > 
> > 
> > [ARMBlockingComment] you also need to support individual gets on each of these Model resources
> > Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:915 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)
> 
> Suhas, we would like your suggestions here. Here is the full story, we originally has a "locations/models" API which return a list of models with self-defined metadata style schema: no "id", "type", "name" and "properties", from this version, we are adding a subscription wide "models" API which use standard ARM resource schema with "id", "type", "name" and "properties". We did add get by Id originally with the new schema, but the tool complain that the schema must be the same as list location models API, that's why we skip adding get by id.



> > > ```
> > >           "$ref": "#/definitions/ModelListResult"
> > > ```
> > 
> > 
> > [ARMBlockingComment] you also need to support individual gets on each of these Model resources
> > Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:915 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)
> 
> Suhas, we would like your suggestions here. Here is the full story, we originally has a "locations/models" API which return a list of models with self-defined metadata style schema: no "id", "type", "name" and "properties", from this version, we are adding a subscription wide "models" API which use standard ARM resource schema with "id", "type", "name" and "properties". We did add get by Id originally with the new schema, but the tool complain that the schema must be the same as list location models API, that's why we skip adding get by id.

I can just explain the ARM requirements here cause there are a few issues. Every resource in ARM must have the id\name\type property. This is used for RBAC and routing by the platform. For location based resources like model this is still true. 

Currently your model properties are added at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail

raosuhas commented:
> > [ARMBlockingComment] you also need to support individual gets on each of these Model resources
> 
> Suhas, these models are base models maintained by the system and not created by customers, it should not be tracked resource type.

They cannot have location in the property bag if they are not a tracked resource. These need to then be proxy arm resources.

raosuhas commented:
Since there are  a few questions around modelling , please book a slot in the ARM modelling office hours and we can discuss it there.

tianxin-ms committed:
using proxy resource type

birdemily commented:
> > > > ```
> > > >   "post": {
> > > > ```
> > > 
> > > 
> > > Why do you need a separate POST action to update the block list items ?
> > > Ifthe updates are part of the body then you should be able to model it as PUT\PATCH , else you can child resources for items instead.
> > > Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:2665 in [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029). [](commit_id = [5a29fea](https://github.com/Azure/azure-rest-api-specs/commit/5a29fea34bf9f8e7436a7f52dae7fa8c82000029), deletion_comment = False)
> > 
> > 
> > These 2 new RaiBlocklistItems APIs are for batch operation (add and delete)
> 
> In ARM we always prefer to model things as resources even items since all tooks like templates \ policy etc depend on the items being modelled as resources. SO for example if you model this as a POST you will not be able to create this as part of a template.

Hi, the requirement comes from customers. They may have thousands raiblocklistitems. It is hard for them to add or delete them 1 by 1.   In this case, we add batch add/delete raiblocklistitems. We also plan to depercate the the single add/delete operations in further if they are not used anymore.

xiaomin-ms commented:
> at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail

We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?

ramoka178 commented:
> > at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail
> 
> We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?

Breaking changes will be reviewed and approved accordingly. Could you tell why you think its not possible?

tianxin-ms commented:
> > > at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail
> > 
> > 
> > We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?
> 
> Breaking changes will be reviewed and approved accordingly. Could you tell why you think its not possible?

Hi @ramoka178 , we joined a ARM modelling office hour Tuesday (4/23) and explained 2 issues raised by last review. 
1. about the "separate POST actions to update the block list items" , this requirement comes from customers, There may be thousands RaiBlockListItems in a RaiBlockList. It is hard for them to add or delete them 1 by 1 using PUT or Delete. These 2 newly added POST APIs are batch operations for this requirement.
2. The newly added models API, we were asked by @raosuhas  to use the tracked resource type and enabled APIs like get single model. However, we already have location-based models API, and this is more like SKU API, the "location" here is like system data not user data, so we don't use tracked resource here. 
Could you help us to address the current issue here? And about the LintDiff error are from old API version.

tianxin-ms committed:
Fix example

tianxin-ms committed:
Remove models API from 04-01-preview

tianxin-ms committed:
Remove sibling values alongside $ref

tianxin-ms commented:
Hi @raosuhas and @raosuhas , we decided to remove the models API from this version, could you please help to review this PR again? We're appreciate your help.

tianxin-ms committed:
Address ModelValidation check

ramoka178 commented:
> > > > at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail
> > > 
> > > 
> > > We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?
> > 
> > 
> > Breaking changes will be reviewed and approved accordingly. Could you tell why you think its not possible?
> 
> Hi @ramoka178 , we joined a ARM modelling office hour Tuesday (4/23) and explained 2 issues raised by last review.
> 
> 1. about the "separate POST actions to update the block list items" , this requirement comes from customers, There may be thousands RaiBlockListItems in a RaiBlockList. It is hard for them to add or delete them 1 by 1 using PUT or Delete. These 2 newly added POST APIs are batch operations for this requirement.
> 2. The newly added models API, we were asked by @raosuhas  to use the tracked resource type and enabled APIs like get single model. However, we already have location-based models API, and this is more like SKU API, the "location" here is like system data not user data, so we don't use tracked resource here.
> 
> About the LintDiff error are from old API version.
> 
> Could you help us to address the current issue here, and when can we get the feedback from the office hour?

For batch APIs, could you refer https://eng.ms/docs/products/arm/api_contracts/batchresourceproviderapi and model your APIs.

tianxin-ms commented:
> > > > > at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail
> > > > 
> > > > 
> > > > We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?
> > > 
> > > 
> > > Breaking changes will be reviewed and approved accordingly. Could you tell why you think its not possible?
> > 
> > 
> > Hi @ramoka178 , we joined a ARM modelling office hour Tuesday (4/23) and explained 2 issues raised by last review.
> > 
> > 1. about the "separate POST actions to update the block list items" , this requirement comes from customers, There may be thousands RaiBlockListItems in a RaiBlockList. It is hard for them to add or delete them 1 by 1 using PUT or Delete. These 2 newly added POST APIs are batch operations for this requirement.
> > 2. The newly added models API, we were asked by @raosuhas  to use the tracked resource type and enabled APIs like get single model. However, we already have location-based models API, and this is more like SKU API, the "location" here is like system data not user data, so we don't use tracked resource here.
> > 
> > About the LintDiff error are from old API version.
> > Could you help us to address the current issue here, and when can we get the feedback from the office hour?
> 
> For batch APIs, could you refer https://eng.ms/docs/products/arm/api_contracts/batchresourceproviderapi and model your APIs.

Hi @ramoka178 , the batch API here means we are adding 2 POST APIs for batch add and batch delete, not to run batch APIs, could you please confirm that can we do like the following?
 POST "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.CognitiveServices/accounts/{accountName}/raiBlocklists/{raiBlocklistName}/addRaiBlocklistItems"

xiaomin-ms commented:
> > > > > at the top level which is causing failures in the automation. All resource properties must be added within the property bag. Unlike your case where you have made the "model" a top level resource. If this is not clear , then please feel free to setup 30 mins in the ARM modelling office hours to discuss in more detail
> > > > 
> > > > 
> > > > We would like to update response schema for the location based list model API, but it would be a breaking change, it seems not possible?
> > > 
> > > 
> > > Breaking changes will be reviewed and approved accordingly. Could you tell why you think its not possible?
> > 
> > 
> > Hi @ramoka178 , we joined a ARM modelling office hour Tuesday (4/23) and explained 2 issues raised by last review.
> > 
> > 1. about the "separate POST actions to update the block list items" , this requirement comes from customers, There may be thousands RaiBlockListItems in a RaiBlockList. It is hard for them to add or delete them 1 by 1 using PUT or Delete. These 2 newly added POST APIs are batch operations for this requirement.
> > 2. The newly added models API, we were asked by @raosuhas  to use the tracked resource type and enabled APIs like get single model. However, we already have location-based models API, and this is more like SKU API, the "location" here is like system data not user data, so we don't use tracked resource here.
> > 
> > About the LintDiff error are from old API version.
> > Could you help us to address the current issue here, and when can we get the feedback from the office hour?
> 
> For batch APIs, could you refer https://eng.ms/docs/products/arm/api_contracts/batchresourceproviderapi and model your APIs.

@ramoka178 we know that API is generic ARM batch solution, but underlines still be multiple requests to RP, which is very useful from client perspective, but it's not as good as the special designed RP batch API: 
- Our batch API allow a single request to add multiple items. 
- because of single requests, we can avoid concurrency issue caused by multiple requests since the parent resource is shared. 
Hope it make sense. Since the RaiBlockListItems is proxy only resource type, I assume we are not limited to only allow "PUT" to create the resource.

TimLovellSmith commented:
please fix linter warnings for examples such as

>⚠️ UniqueXmsExample	Do not have duplicate name of x-ms-example, make sure every x-ms-example name unique. Duplicate x-ms-example: GetDefenderForAISetting
Location: Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json#L3473

TimLovellSmith commented:
Why not just use a single model type for your result lists, instead of two? These look semantically compatible, and I think they may as well both support paging.

>
```   "RaiBlockListItemsResult": {
      "type": "object",
      "properties": {
        "nextLink": {
          "description": "The link used to get the next page of RaiBlocklistItems.",
          "type": "string"
        },
        "value": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/RaiBlocklistItem"
          },
          "x-ms-identifiers": [
            "name"
          ],
          "description": "The list of RaiBlocklistItems."
        }
      },
      "description": "The list of cognitive services RAI Blocklist Items."
    },
    "RaiBlocklistItems": {
      "description": "The list of Cognitive Services RaiBlocklist Items.",
      "type": "array",
      "items": {
        "$ref": "#/definitions/RaiBlocklistItem"
      },
      "x-ms-identifiers": [
        "name"
      ]
    },
```

TimLovellSmith commented:
For your request payload type Definitions for bulk operations, how would you feel about some name changes? I think these would be easier to understand:

RaiBlocklistItemsAddRequestBody -> RaiBlocklistItemsBulkAddRequest
RaiBlocklistItemsDeleteRequestBody -> RaiBlocklistItemsBulkDeleteRequest

etc.

tianxin-ms committed:
Address comments

tianxin-ms committed:
fix breaking change

tianxin-ms commented:
Hi @TimLovellSmith, I've fixed linter warnings for examples, and removed duplicate model "RaiBlocklistItems". and update the request payload name as suggestion. I'll be greatly appreciated it if you could take a look at it again, it's been a long time for this PR review.

tianxin-ms commented:
@TimLovellSmith , thanks for the signoff, but seems you also added ARMChangesRequested, what do we need to change? Could you also help with Lintdiff and sdk-for-go? the Lintdiff error are from last versions.

TimLovellSmith commented:
Something weird is happening with labels, sorry..

TimLovellSmith commented:
I think it'd also be great to have NetworkSecurityPerimeter definitions in their own separate file: NetworkSecurityPerimeter.json similar to what BotService has here:

https://github.com/Azure/azure-rest-api-specs/blob/874c4f1c7a71be5357d92e4c6b9220bd91e77741/specification/botservice/resource-manager/Microsoft.BotService/preview/2023-09-15-preview/networksecurityperimeter.json

I should probably do some more consistency checks between those...

TimLovellSmith commented:
>           "$ref": "#/definitions/NetworkSecurityPerimeterAccessRulePropeties"

theres a typo here:

NetworkSecurityPerimeterAccessRuleProperties

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6489 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

I don't see some other services using direction in their definition, I wonder why...?

Also, is it an either-or-choice?

What are the valid values? Can you define them in an enum?

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6500 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>     "NetworkSecurityPerimeterConfigurationListResult": {

Would suggest calling it 'NetworkSecurityPerimeterConfigurationList' for consistency with other RPs

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6317 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

I think you should define a string enum with the actual ProvisioningState values that you would return, for the nsp configuration. Other RPs are doing so.

e.g.

```

        "provisioningState": {
          "enum": [
            "Creating",
            "Updating",
            "Accepted",
            "Succeeded",
            "Failed",
            "Deleting"
          ],
          "type": "string",
          "x-ms-enum": {
            "name": "provisioningState",
            "modelAsString": true
          },
          "default": "Succeeded"
        },
```

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6360 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>     "NSPProvisioningIssue": {

Calling it just ProvisioningIssue would be more consistent with the other RPs I'm looking at

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6382 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

Maybe you should use what other RPs have defined - would this list be complete for you?:

```
   "severity": {
              "description": "Severity of the error"
              "enum": [
                "Warning",
                "Error"
              ],
              "type": "string",
              "x-ms-enum": {
                "name": "Severity",
                "modelAsString": true
              }
            },
```

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6405 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

Compare to this similar definition which might be better

```
            "suggestedResourceIds": {
              "type": "array",
              "items": {
                "readOnly": true,
                "type": "string",
                "format": "arm-id",
                "description": "Fully qualified resource ID for the resource. E.g. \"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}\""
              },
              "readOnly": true,
              "description": "IDs of resources that can be associated to the same perimeter to remediate the issue."
            },
```

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6416 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         }

Consider changing to

```
 "accessMode": {
          "description": "Access Mode of the resource association",
          "enum": [
            "Enforced",
            "Learning",
            "Audit"
          ],
          "type": "string",
          "x-ms-enum": {
            "name": "AccessMode",
            "modelAsString": true
          }
        }
```

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6455 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>       "description": "Network Security Perimeter data",

I'd suggest

      "description": "Information about a linked Network Security Perimeter",

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6427 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>           "type": "string"


          "format": "arm-id",

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6432 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

this should be a numeric type

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6469 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>     "NetworkSecurityPerimeterProfileInfo": {

You won't support diagnosticSettingsVersion or enabledLogCategories properties yet?

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6458 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

TimLovellSmith commented:
>         },

Maybe I was overlooking their definition. I see some define it as an enum

```

            "direction": {
              "description": "Direction of Access Rule",
              "enum": [
                "Inbound",
                "Outbound"
              ],
              "type": "string",
              "x-ms-enum": {
                "name": "NspAccessRuleDirection",
                "modelAsString": true
              }
            },
```

---
In reply to: [2087770394](https://github.com/Azure/azure-rest-api-specs/pull/28509#issuecomment-2087770394) [](http://example.com/codeflow?ancestors=2087770394)

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6500 in ef26cc1. [](commit_id = ef26cc1859b0450d2480482b77ce9e01e941cf80, deletion_comment = False)

tianxin-ms committed:
address NSP comments

tianxin-ms committed:
fix nsp example

tianxin-ms committed:
fix NSP

tianxin-ms commented:
/pr RequestMerge

tianxin-ms committed:
fix an example

TimLovellSmith commented:
azure-sdk-for-go is blocking merge, please address then request merge again

tianxin-ms committed:
fix sdk-go

tianxin-ms commented:
It's fixed, thank you so much for your help. @TimLovellSmith

tianxin-ms commented:
/pr RequestMerge

TimLovellSmith commented:
@tianxin-ms I'm working on adding NetworkSecurityPerimeters to common-types, as it seems clearly needed! Please could you help me review the PR? Thanks. https://github.com/Azure/azure-rest-api-specs/pull/28958

TimLovellSmith commented on specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:
```diff
@@ -0,0 +1,6791 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.CognitiveServices/locations/{location}/modelCapacities/{model-format}/{model-name}/{model-version}": {
```
[ARMBlockingComment] Unfortunately I just noticed this and I don't think it follows ARM guidelines - ARM (and RBAC) needs the type/{instance}/type/{instance}/type/{instance} pattern to be followed in resource URLs

tianxin-ms committed:
Fix modelCapacities APIs

TimLovellSmith commented:
@tianxin-ms There are some ModelValidation errors to fix now

TimLovellSmith commented:
>     },

This is basically okay. Now that I think of it further, it might be even better just to have one complex filter parameter in line with the well-known 'list resources' API, which allows OData style filtering by e.g. type+region. But I don't think that is actually required, and it could even be added in a future version, if all the filter parameters are optional

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6771 in a2e20e1. [](commit_id = a2e20e1522525f0b9b3867b0b4e6ecb6a41aec53, deletion_comment = False)

TimLovellSmith commented:
>       "required": true,

Just wondering - do you really wat to make it required? Maybe it would be useful to not need to specify all the parameters, and be able to query different subsets?

---
Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6766 in a2e20e1. [](commit_id = a2e20e1522525f0b9b3867b0b4e6ecb6a41aec53, deletion_comment = False)

tianxin-ms committed:
removed $

tianxin-ms commented:
> @tianxin-ms There are some ModelValidation errors to fix now

Thanks, I just fixed it, a typo

tianxin-ms commented:
> > ```
> >   "required": true,
> > ```
> 
> Just wondering - non-blocking -- do you really wish to make it required? Maybe it would be useful to not need to specify all the parameters, and be able to query different subsets?
> 
> Refers to: specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/preview/2024-04-01-preview/cognitiveservices.json:6766 in [0a714ec](https://github.com/Azure/azure-rest-api-specs/commit/0a714ec6ac34248c5bd0597e117235121dfffe57). [](commit_id = [0a714ec](https://github.com/Azure/azure-rest-api-specs/commit/0a714ec6ac34248c5bd0597e117235121dfffe57), deletion_comment = False)

It's designed to be required, those 3 are all needed to query a "model". We will return bad request, if every of them is missing.

tianxin-ms commented:
@TimLovellSmith , the ModelValidation error is fixed now, could you please help us signoff and approve the manifest PR?

slreznit commented:
> @TimLovellSmith , the ModelValidation error is fixed now, could you please help us signoff and approve the manifest PR?

+1 We are waiting for this manifest for over a month now and this is blocking our product release

madhura-np committed:
Merge branch 'main' into tianxinliu/2024-04-01-preview

madhura-np committed:
Merge branch 'main' into tianxinliu/2024-04-01-preview
